File: services/executor-runtime/bun/cgroup_limits.ts
/**
 * Linux cgroup v2 helpers for per-snippet CPU limits. No-op on non-Linux.
 */

import { existsSync, mkdirSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { join } from "node:path";

const CGROUP_ROOT = "/sys/fs/cgroup";

function linuxCgroupAvailable(): boolean {
  return process.platform === "linux" && existsSync(CGROUP_ROOT);
}

export function applyLimits(
  pid: number,
  maxMemoryMb: number,
  maxCpuPercent: number,
): string | null {
  if (!linuxCgroupAvailable()) return null;
  if (maxMemoryMb <= 0 && maxCpuPercent <= 0) return null;

  const cg = join(CGROUP_ROOT, `velane-${pid}`);
  try {
    mkdirSync(cg, { recursive: true });
    if (maxMemoryMb > 0) {
      writeFileSync(join(cg, "memory.max"), String(maxMemoryMb * 1024 * 1024));
    }
    if (maxCpuPercent > 0 && maxCpuPercent <= 100) {
      const period = 100000;
      const quota = Math.max(1, Math.floor((period * maxCpuPercent) / 100));
      writeFileSync(join(cg, "cpu.max"), `${quota} ${period}`);
    }
    writeFileSync(join(cg, "cgroup.procs"), String(pid));
    return cg;
  } catch {
    return null;
  }
}

export function applyCpuLimit(pid: number, maxCpuPercent: number): string | null {
  return applyLimits(pid, 0, maxCpuPercent);
}

export function applyMemoryLimitCgroup(pid: number, maxMemoryMb: number): string | null {
  return applyLimits(pid, maxMemoryMb, 0);
}

export function readPeakMemoryMb(cgroup: string | null): number {
  if (!cgroup) return 0;
  try {
    const raw = readFileSync(join(cgroup, "memory.peak"), "utf8").trim();
    if (!raw || raw === "max") return 0;
    return Math.floor(Number(raw) / (1024 * 1024));
  } catch {
    return 0;
  }
}

export function cleanupCgroup(cgroup: string | null): void {
  if (!cgroup) return;
  try {
    rmSync(cgroup, { recursive: true, force: true });
  } catch {
    // ignore
  }
}

export function isOomExit(exitCode: number): boolean {
  return exitCode === -9 || exitCode === 137;
}

File: services/executor-runtime/python/cgroup_limits.py
"""Linux cgroup v2 helpers for per-snippet CPU limits. No-op on non-Linux."""

from __future__ import annotations

import platform
from pathlib import Path

_CGROUP_ROOT = Path("/sys/fs/cgroup")


def _linux_cgroup_available() -> bool:
    return platform.system() == "Linux" and _CGROUP_ROOT.is_dir()


def apply_limits(pid: int, max_memory_mb: int, max_cpu_percent: int) -> Path | None:
    """Apply memory.max and cpu.max in a single ephemeral cgroup."""
    if not _linux_cgroup_available():
        return None
    if max_memory_mb <= 0 and max_cpu_percent <= 0:
        return None

    cg = _CGROUP_ROOT / f"velane-{pid}"
    try:
        cg.mkdir(exist_ok=True)
        if max_memory_mb > 0:
            (cg / "memory.max").write_text(str(max_memory_mb * 1024 * 1024))
        if max_cpu_percent > 0 and max_cpu_percent <= 100:
            period = 100000
            quota = max(1, int(period * max_cpu_percent / 100))
            (cg / "cpu.max").write_text(f"{quota} {period}")
        (cg / "cgroup.procs").write_text(str(pid))
        return cg
    except OSError:
        return None


def apply_cpu_limit(pid: int, max_cpu_percent: int) -> Path | None:
    return apply_limits(pid, 0, max_cpu_percent)


def apply_memory_limit_cgroup(pid: int, max_memory_mb: int) -> Path | None:
    return apply_limits(pid, max_memory_mb, 0)


def read_peak_memory_mb(cgroup: Path | None) -> int:
    if cgroup is None:
        return 0
    peak = cgroup / "memory.peak"
    try:
        raw = peak.read_text().strip()
        if not raw or raw == "max":
            return 0
        return int(int(raw) / (1024 * 1024))
    except OSError:
        return 0


def cleanup_cgroup(cgroup: Path | None) -> None:
    if cgroup is None:
        return
    try:
        cgroup.rmdir()
    except OSError:
        pass


def is_oom_exit(exit_code: int | None) -> bool:
    # SIGKILL (-9) often indicates cgroup OOM killer.
    return exit_code == -9 or exit_code == 137
